### ISSUE TITLE
"Please hold" landing pages for slow scale from zero scenarios

### ISSUE THREAD

**Issue by worldspawn** [kedacore/http-add-on#874]
### Proposal

If the service is scaling from zero and maybe needs to wait for a new node to provision a nicer user XP might be to just land on a page that says "Teamcity is starting up... please wait". Could be enabled and configured in the http scaled object resource. MVP could be a very simple page saying "{SomeValueFromHttpResource} is start up please wait" and it polls.

### Use-Case

Nodes can be slow (well a cpl of minutes) to start up. The experience scaling from zero in this case is pretty ordinary. The default wait time is only 30s and then you Bad Gateway or something like that.

I nice web page telling people to sit tight and wait for the magic would be a lot nicer.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by JorTurFer** [comment]
Hum...
It's an interesting feature, but I'm not sure about how we can implement this. I mean, a simple message it's nice, but I'm afraid about if this has to scale to serve complex HTML and how we can do this.
WDYT @tomkerkhove ?
Maybe we can allow custom messages for HTTP Errors somehow

**Comment by thincal** [comment]
I come up with similar requirement, just customizing the error message is enough for me.

**Comment by JorTurFer** [comment]
@tomkerkhove ?

**Comment by lenglet-k** [comment]
i'm interesting too for this feature. Maybe have a light custom HTML by httpsScaleObject. For not afraid user, just only print a loading message.

**Comment by zroubalik** [comment]
Yeah, this should be imho configurable. And serve the light custom HTML for users that want to. They should be able to provide their custom HTLM file.

**Comment by JorTurFer** [comment]
make sense, but maybe I'd add another CRD for this, as it's something that can be reusable (or not). WDYT?

**Comment by thincal** [comment]
> by content and return code

+1

**Comment by thincal** [comment]
Any progress here

**Comment by JorTurFer** [comment]
No, currently there isn't anybody working on this, are you willing to tackle it?

**Comment by thincal** [comment]
> No, currently there isn't anybody working on this, are you willing to tackle it?

Yes, I could have a try. My quick idea is to add the response content and code into existing http-scaled-object, but it seems that there is suggestion about using a new CRD, could help explain the consideration behind?

**Comment by JorTurFer** [comment]
Yeah, I suggested the new CRD because I can imagine a scenario where you have multiple services behind the proxy, and you want to provide a custom message for all of them. 
The new CRD could be just something like:
```
spec:
  httpStatus: XXX (number)
  html: |
    <html>....</html>
```

and then, we can include in the HTTPScaledObject with something like:
```
spec:
  loadingTemplate: xxxxxxx
  minReplicaCount: 0
  ....
```

Currently, these templates can be cluster scoped and maybe in the future we can support both, clustered and namespaced.
This will keep the HTTPScaledObject cleaner because if you have to introduce the HTML within the HTTPScaledObject, it can introduce a lot of new lines (even more when you want to reuse them)

**Comment by calvinbui** [comment]
Sablier has this feature: https://acouvreur.github.io/sablier/#/strategies

**Comment by kahirokunn** [comment]
It would be great if we could automatically inject JavaScript into the HTML to implement polling and have the feature automatically access the HTTP server as soon as it becomes available.

**Comment by julopezc-acciona** [comment]
I have a annotation using https://github.com/kubernetes/ingress-nginx/ which allows to handle custom http errors:
```yaml
    nginx.ingress.kubernetes.io/custom-http-errors: "502"
    nginx.ingress.kubernetes.io/default-backend: keda-autoscale-502-error-svc
```

`keda-autoscale-502-error-svc` is a service that lives in the same namespace that keda is installed and has a simple nginx deployment with a default.conf that reloads the current page:
```conf
    server {
        listen 80;
        location / {

            add_header Content-Type text/html;
            return 200 '<!DOCTYPE html>
            <html>
            <head>
                <title>Application Scaling</title>
                <style>
                    body {
                        font-family: Arial, sans-serif;
                        text-align: center;
                        padding-top: 50px;
                        background-color: #f0f0f0;
                    }
                    h2 {
                        color: #333;
                    }
                    p {
                        color: #666;
                    }
                </style>
                <script>
                    setTimeout(function() {
                        window.location.reload();
                    }, 5);
                </script>
            </head>
            <body>
                <h2>Application $host scaling up.</h2>
                <p>Please wait, this page will automatically refresh...</p>
            </body>
            </html>';
        }
    }

```

**Comment by julopezc-acciona** [comment]
Yes, I do have a ingress object in keda's namespace, then it routes the incoming traffic to the relevant service in another namespace

**Comment by florianbeer** [comment]
Then this probably won't work for my use case. I'd need one central "wake up" page for multiple workloads with ingresses in their own namespaces.

But very nice solution nevertheless 👍

**Comment by kahirokunn** [comment]
Keep

**Comment by yyewolf** [comment]
Hi everyone, concerning this issue (and maybe other use cases?), would it be maybe possible to consider having a `fallbackService` in the `HTTPScaledObject` CRD ?

This would be an optional field that when present, would redirect traffic to this service when the Deployment is either not scaled to zero or not ready.

**Comment by j-cull** [comment]
When is this likely to be released?

**Comment by linkvt** [comment]
Hi @blkperl ,

I created a new PR to implement the landing page/placeholder feature based on the existing work in the other PR, see #1626 

About how that would solve your ideal requirements:
The ideal waiting/landing page would include
1. the ability to show real time progress of pods launching (or not launching)
	- we don't have this info right now and can't display it, scaling happens outside of the HTTP Addon
	- 👉 won't be implemented in the foreseeable future
2. automatic refresh with a redirect once the service is ready
	- could be implemented with:
		- the [`Refresh` header](https://developer.mozilla.org/de/docs/Web/HTTP/Reference/Headers/Refresh) set to a few seconds
		- your custom response containing the [meta http-equiv tag](https://developer.mozilla.org/de/docs/Web/HTML/Reference/Elements/meta/http-equiv)
		- your custom response including some inline JS fetching the current URL to check when no placeholder is returned anymore and then do the full page reload to get to the actual app
3. estimated time to completion
	- same as with 1., scaling information is a) outside of the HTTP Addon and also impossible to know as it heavily depends on your environment
	- 👉 we will never be able to calculate something like this, you would need to specify your expected wait time based on your enviroment
4. links to get help
	- can be implemented in the custom page you can provide
5. custom html/css to align to our branding or customer branding requirements
	- same as 4.
	- you could share different placeholder pages among InterceptorRoutes
	- templating isn't included yet

Let me know what you think
